src/gui/widgets/typed_debugger.h

```diff
@@ -0,0 +1,160 @@
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ *   This program is distributed in the hope that it will be useful,       *
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
+ *   GNU General Public License for more details.                          *
+ *                                                                         *
+ *   You should have received a copy of the GNU General Public License     *
+ *   along with this program; if not, write to the                         *
+ *   Free Software Foundation, Inc.,                                       *
+ *   51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.           *
+ ***************************************************************************/
+
+#pragma once
+
+#include <string>
+#include <unordered_map>
+#include <variant>
+#include <vector>
+
+#include "core/debug.h"
+#include "gui/widgets/filedialog.h"
+#include "imgui.h"
+#include "support/eventbus.h"
+
+// Represents the type-name-size tuple for either a struct field or a function argument.
+struct FieldOrArgumentData {
+    std::string type;
+    std::string name;
+    size_t size;
+};
+
+// Information about a read/write instruction, including the name of the function from which it was emitted.
+struct ReadWriteLogEntry {
+    enum class AccessType { Read, Write };
+
+    uint32_t instructionAddress;
+    std::string functionName;
+    AccessType accessType;
+};
+
+// A WatchTreeNode contains the information necessary for displaying a type hierarchy in a tree.
+struct WatchTreeNode {
+    std::string type;
+    std::string name;
+    size_t size = 0;
+    std::vector<WatchTreeNode> children;
+    std::vector<ReadWriteLogEntry> logEntries;
+};
+
+// For an explanation of the meaning of addressOfPointer, see declaration of displayNode().
+struct AddressNodeTuple {
+    uint32_t address;
+    bool addressOfPointer;
+    WatchTreeNode node;
+};
+
+struct DebuggerFunction {
+    std::string name;
+    std::vector<FieldOrArgumentData> arguments;
+};
+
+// Represents a value stored in a register, as opposed to one stored at an address in memory.
+struct RegisterValue {
+    uint32_t value;
+    std::string type;
+    std::string name;
+    size_t size = 0;
+};
+
+namespace PCSX {
+class GUI;
+namespace Widgets {
+
+class TypedDebugger {
+  public:
+    void draw(const char* title, GUI* gui);
+    bool& m_show;
+    TypedDebugger(bool& show);
+
+    /**
+     * Data importation.
+     */
+
+    Widgets::FileDialog m_importDataTypesFileDialog = {[]() { return _("Import data types"); }};
+    Widgets::FileDialog m_importFunctionsFileDialog = {[]() { return _("Import functions"); }};
+    enum class ImportType { DataTypes, Functions };
+    void import(const char* filename, const ImportType& importType);
```

Comment:
It might make more sense to take a string_view of the contents of the file, as an argument, instead of the filename, for more flexibility of the import process

Also, taking a reference of an int-like is more expensive than just passing it by value :)

Reply:
Oops, I usually do pass types of that size by value when it's semantically appropriate. :)